[WIP] Reason about iterating over a map's key set using an enhanced for loop

msridhar opened the pull request:
Fixes #535 

Uses the strategy outlined in this comment: https://github.com/uber/NullAway/issues/535#issuecomment-1006811236

Still a WIP as it needs more tests

lazaroclapp commented on nullaway/src/test/java/com/uber/nullaway/NullAwayKeySetIteratorTests.java:
```diff
@@ -0,0 +1,132 @@
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import java.util.*;",
```
I know this is test code, so arguably this doesn't matter, but shouldn't this be just `import java.util.Map`?

lazaroclapp commented on nullaway/src/test/java/com/uber/nullaway/NullAwayKeySetIteratorTests.java:
```diff
@@ -0,0 +1,132 @@
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import java.util.*;",
```
As above.

lazaroclapp commented on nullaway/src/test/java/com/uber/nullaway/NullAwayKeySetIteratorTests.java:
```diff
@@ -0,0 +1,132 @@
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import java.util.*;",
```
Even here, I'd just import `Map` and `HashMap`.

lazaroclapp replied:
Just, in general, mild vote for listing the imports individually as if this were real code. But I am open to the opposite argument.

msridhar replied:
Agreed, fixed in 4257191

lazaroclapp commented on nullaway/src/test/java/com/uber/nullaway/NullAwayKeySetIteratorTests.java:
```diff
@@ -0,0 +1,132 @@
+            "public class Test {",
+            "  public void keySetStuff(Map<Object, Object> m) {",
+            "    // BUG: Diagnostic contains: dereferenced expression",
+            "    m.get(m.keySet().iterator().next()).toString();",
```
Is this unhandled or a true positive? Technically the map could be empty....

msridhar replied:
If the Map were empty, this would cause a `NoSuchElementException`, not an NPE.  So I would say our report is a false positive, since it refers to the possibility of an NPE.

lazaroclapp commented on nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPath.java:
```diff
@@ -658,6 +693,46 @@ public boolean equals(Object obj) {
+     * as this class is designed specifically for reasoning about iterating over map keys using an
+     * enhanced-for loop over a {@code keySet()}, and for such cases the iterator is always stored
+     * locally
+     */
```
If we only support locals, is there a more precise type than `Element`?

msridhar replied:
Yup, tightened to `VariableElement` in 76ddc4c

lazaroclapp commented on nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java:
```diff
@@ -530,6 +534,94 @@ private void handleEqualityComparison(
+   * {@code null}.
+   */
+  @Nullable
+  private Node getMapNodeForKeySetIteratorCall(MethodInvocationNode invocationNode) {
```
Worth a preconditions check that the method being called is `Set::iterator()`? Or do worry about the extra performance hit of those checks and feel the `isEnhancedForIteratorVariable(...)` check on the caller and our assumptions about CF's CFG generation are enough of a check?

msridhar replied:
In general, I was trying to minimize the overhead from unnecessary checks.  I am very confident in the invariants we rely on in terms of the CF CFG right now.  My hope was that if we bump the CF version and something changes, our tests would fail.  I guess there is the possibility that CF would change in a way that would cause us to start thinking other types of expressions were non-null erroneously, introducing a new unsoundness.  That seems pretty unlikely to me, but if you'd like, I can introduce some extra sanity checking.  I would want to limit that checking to not run for every assignment statement, though, instead only for cases where we are pretty sure we're in an enhanced for loop. WDYT?

lazaroclapp replied:
Your call. I think you are way more familiar with the CF internals than I am, if you feel the invariant is pretty much set in stone and requires no runtime checking, I am fine with that. That said, if you think there is a reasonable value in adding sanity checking, I definitely agree we can defer it until right before the `updates.set(...)` call. Though, again, if you think our assumptions are unlikely to break, I am also fine leaving this as is.

lazaroclapp commented on nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java:
```diff
@@ -530,6 +534,94 @@ private void handleEqualityComparison(
+          updates.set(mapWithIteratorContentsKey, NONNULL);
+        }
+      }
+    } else {
```
Why `else { if {...} }` instead of `else if { ... }`?

msridhar replied:
Good catch, fixed in e628368

lazaroclapp commented on nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java:
```diff
@@ -530,6 +534,94 @@ private void handleEqualityComparison(
+    } else {
+      // Check for an assignment lhs = iter#numX.next().  From the structure of Checker Framework
+      // CFGs, we know that if iter#numX is the receiver of a call on the rhs of an assignment, it
+      // must be a call to next().
```
Wonder if we should trust that CFG is stable enough here, or if we should be checking the method symbol for `Iterator::next()`. We can do this after verifying `isEnhancedForIteratorVariable(...)` but before generating the AP, if we are worried about the performance hit from doing so on every method invocation.

msridhar replied:
See [my comment above](https://github.com/uber/NullAway/pull/554#discussion_r789846244).  If you'd like a check here, is it ok to do it only if `mapGetPath != null`?  That would limit the perf impact even more (we will likely have many enhanced-for loops that are not over the key set of a map).

lazaroclapp commented on nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPath.java:
```diff
@@ -468,6 +468,36 @@ private static Root populateElementsRec(
+   * Creates an access path identical to {@code accessPath} (which must represent a map), but with
+   * {@code mapKey} as its map {@code get()} argument
+   */
+  public static AccessPath withMapKey(AccessPath accessPath, MapKey mapKey) {
```
So, the way we are using this method is that we have an AP of the form `ap1 = x.y.m.get(iter#numX)` (which is not real, more of a virtual map get) and see `v = iter#numX.next()` so we call this method with `(ap1, v)` to obtain `ap1 = x.y.m.get(v)`. Correct?

If so, maybe this should be `replaceMapKey(...)`? Or, if that's still missing some usage where this is called an access path without an existing map key, then the javadoc should mention it nonetheless "{@code mapKey} as its map {@code get()} argument, replacing any map key information in the original {@code accessPath}."

lazaroclapp reviewed (commented):
A few comments and questions, but overall looks good! 

Did we run our performance benchmarks on this change?

msridhar commented:
> Did we run our performance benchmarks on this change?

No, not yet, I'll do that now.

Thanks for the review @lazaroclapp!  I addressed most of the comments.  There is still an open question about how many well-formedness checks we should add on the CFG; see my inline comments.

lazaroclapp reviewed (commented):
See comment below. I'd be fine with this landing as is after performance benchmarking. I leave adding well-formedness checks up to what you think makes sense. If you are pretty sure the CFG invariant will not change in an unexpected way for future Java constructs, then it makes sense to omit those checks for performance reasons (but having them and deferring them to the last possible moment is always an option).

msridhar commented:
Here are the perf benchmarking results.  master branch:
```
Benchmark                      Mode  Cnt   Score   Error  Units
AutodisposeBenchmark.compile  thrpt   25  10.837 ± 0.078  ops/s
CaffeineBenchmark.compile     thrpt   25   2.390 ± 0.013  ops/s
```

With this PR:
```
Benchmark                      Mode  Cnt   Score   Error  Units
AutodisposeBenchmark.compile  thrpt   25  10.851 ± 0.070  ops/s
CaffeineBenchmark.compile     thrpt   25   2.362 ± 0.019  ops/s
```

The changes seem to be within the noise.

msridhar commented:
I decided to add the sanity checks, just to be safe.  Will re-run perf benchmarks now.  If the checks do end up causing overhead we care about (based on profiling), we can remove them later (or find something else to optimize).

lazaroclapp commented on nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java:
```diff
@@ -601,17 +611,22 @@ private Node getMapNodeForKeySetIteratorCall(MethodInvocationNode invocationNode
     return null;
   }
 
+  private boolean isCallToMethod(
```
👍

lazaroclapp reviewed (approved):
Approved iff no performance regression :)

msridhar commented:
> Approved iff no performance regression :)

Re-ran and perf differences are within noise.  Landing 🙂
